aws/redirect/main.go

```
@@ -2,30 +2,81 @@ package main
+	"github.com/aws/aws-sdk-go-v2/service/lambda/types"
 )
 
-func handler(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
-	slog.Debug("Handling redirect", "RawPath", req.RawPath, "param[key]", req.PathParameters[service.IdParam])
-	status, url, errStr := service.RetrieveValueAndMarshalError(ctx, shorturl.RetrieveSQL, req.PathParameters[service.IdParam])
+var lambdaClient *lambdasdk.Client
+
+func init() {
+	cfg, err := config.LoadDefaultConfig(context.Background())
+	if err != nil {
+		panic("failed to load AWS config: " + err.Error())
+	}
+	lambdaClient = lambdasdk.NewFromConfig(cfg)
+}
+
+func redirect(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
+	key := req.PathParameters[service.IdParam]
+	slog.Info("Handling redirect", "RawPath", req.RawPath, "param[key]", key)
+	status, url, errStr := service.RetrieveValueAndMarshalError(ctx, shorturl.RetrieveSQL, key)
 	if errStr != "" {
 		return events.APIGatewayV2HTTPResponse{StatusCode: status, Body: errStr}, nil
 	}
 
+	sendStatistics(ctx, key, req)
+
 	return events.APIGatewayV2HTTPResponse{
 		StatusCode: http.StatusMovedPermanently, // TODO: in future we can return 302 if the URL TTL is short or 307 or  308 if we will support methods other then GET
 		Headers: map[string]string{
-			"Location": url.Target,
+			"Location":      url.Target,
+			"Cache-Control": "no-store, no-cache, must-revalidate, max-age=0",
 		},
 	}, nil
 }
 
+func sendStatistics(ctx context.Context, key string, req events.APIGatewayV2HTTPRequest) {
+	event := stats.StatsEvent{
+		Key:       key,
+		IP:        req.RequestContext.HTTP.SourceIP,
+		UserAgent: req.Headers["user-agent"],
+		Referer:   req.Headers["referer"],
+		Timestamp: time.Now().UTC(),
+		Language:  req.Headers["accept-language"],
+	}
+
+	payload, err := json.Marshal(event)
+	if err != nil {
+		slog.Warn("Failed to marshal stats event", "error", err)
```

It's better to return the wrapped error and handle it on the caller's side, either by logging or returning.